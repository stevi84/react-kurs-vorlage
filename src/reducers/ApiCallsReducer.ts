import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from './Store';

export const apiCallsInitialState = {
  runningReads: 0,
  runningSubmits: 0,
};

const apiCallsSlice = createSlice({
  name: 'apiCalls',
  initialState: apiCallsInitialState,
  reducers: {
    increaseReads(state, _action: PayloadAction) {
      return {
        ...state,
        runningReads: state.runningReads + 1,
      };
    },
    decreaseReads(state, _action: PayloadAction) {
      return {
        ...state,
        runningReads: Math.max(0, state.runningReads - 1),
      };
    },
    increaseSubmits(state, _action: PayloadAction) {
      return {
        ...state,
        runningSubmits: state.runningSubmits + 1,
      };
    },
    decreaseSubmits(state, _action: PayloadAction) {
      return {
        ...state,
        runningSubmits: Math.max(0, state.runningSubmits - 1),
      };
    },
  },
});

const { actions, reducer } = apiCallsSlice;
export const { increaseReads, decreaseReads, increaseSubmits, decreaseSubmits } = actions;
export { reducer as apiCallsReducer };

export const isReadingSelector = (state: RootState): boolean => state.apiCalls.runningReads > 0;
export const isSubmittingSelector = (state: RootState): boolean => state.apiCalls.runningSubmits > 0;
