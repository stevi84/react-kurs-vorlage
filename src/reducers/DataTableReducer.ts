import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from './Store';

export const dataTableInitialState = {
  visible: true,
};

const dataTableSlice = createSlice({
  name: 'dataTable',
  initialState: dataTableInitialState,
  reducers: {
    setDataTableVisible(state, action: PayloadAction<boolean>) {
      state.visible = action.payload;
    },
  },
});

const { actions, reducer } = dataTableSlice;
export const { setDataTableVisible } = actions;
export { reducer as dataTableReducer };

export const dataTableVisibleSelector = (state: RootState): boolean => state.dataTable.visible;
