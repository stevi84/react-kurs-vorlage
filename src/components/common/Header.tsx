import { LanguageSwitch } from './LanguageSwitch';
import './header.css';
import { ChangeEvent, useState } from 'react';
import { useAppDispatch, useAppSelector } from '../../reducers/Store';
import { dataTableVisibleSelector, setDataTableVisible } from '../../reducers/DataTableReducer';

export const Header = () => {
  const dispatch = useAppDispatch();
  const [showLogo, setShowLogo] = useState<boolean>(true);
  const dataTableVisible = useAppSelector(dataTableVisibleSelector);
  const onLogoChange = (e: ChangeEvent<HTMLInputElement>) => setShowLogo(e.target.checked);
  const onTableChange = (e: ChangeEvent<HTMLInputElement>) => dispatch(setDataTableVisible(e.target.checked));

  return (
    <div className='header-container'>
      {showLogo && <div data-testid='header-logo' className='header-item header-icon'>
        üåê
      </div>}
      <div className='header-item header-text'>
        React-App
      </div>
      <div className='header-item'>
        <input id='header-logo-checkbox' type='checkbox' checked={showLogo} onChange={onLogoChange}></input>
        <label htmlFor='header-logo-checkbox'>Logo</label>
      </div>
      <div className='header-item'>
        <input id='header-table-checkbox' type='checkbox' checked={dataTableVisible} onChange={onTableChange}></input>
        <label htmlFor='header-table-checkbox'>Tabelle</label>
      </div>
      <div className='header-item'>
        <LanguageSwitch />
      </div>
    </div>
  );
};
