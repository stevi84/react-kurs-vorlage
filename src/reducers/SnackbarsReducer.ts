import { Action, createReducer } from '@reduxjs/toolkit';
import { OptionsObject, SnackbarKey, SnackbarMessage } from 'notistack';
import { RootState } from './Store';

// https://stackoverflow.com/questions/62754584/how-can-i-use-notistick-any-snackbar-with-redux-toolkit-and-react
// https://codesandbox.io/s/github/iamhosseindhv/notistack/tree/master/examples/redux-example?file=/redux/actions.js:0-150

interface Snackbar {
  message: SnackbarMessage;
  options?: OptionsObject;
  key: SnackbarKey;
  dismissed: boolean;
}

// Actions

const ENQUEUE_SNACKBAR = 'ENQUEUE_SNACKBAR';
const CLOSE_SNACKBAR = 'CLOSE_SNACKBAR';
const REMOVE_SNACKBAR = 'REMOVE_SNACKBAR';

interface EnqueueSnackbarAction extends Action<string> {
  notification: Snackbar;
}

export const enqueueSnackbar = (message: SnackbarMessage, options?: OptionsObject): EnqueueSnackbarAction => {
  return {
    type: ENQUEUE_SNACKBAR,
    notification: {
      message,
      options,
      key: (options && options.key) || new Date().getTime() + Math.random(),
      dismissed: false,
    },
  };
};

interface CloseSnackbarAction extends Action<string> {
  key: SnackbarKey | undefined;
  dismissAll: boolean;
}

export const closeSnackbar = (key?: SnackbarKey): CloseSnackbarAction => ({
  type: CLOSE_SNACKBAR,
  dismissAll: !key, // dismiss all if no key has been defined
  key,
});

interface RemoveSnackbarAction extends Action<string> {
  key: SnackbarKey;
}

export const removeSnackbar = (key: SnackbarKey): RemoveSnackbarAction => ({
  type: REMOVE_SNACKBAR,
  key,
});

// Reducer

interface SnackbarState {
  notifications: Snackbar[];
}

export const snackbarsInitialState: SnackbarState = {
  notifications: [],
};

export const snackbarsReducer = createReducer(snackbarsInitialState, {
  [ENQUEUE_SNACKBAR]: (state, action: EnqueueSnackbarAction) => {
    return {
      notifications: [...state.notifications, { ...action.notification }],
    };
  },
  [CLOSE_SNACKBAR]: (state, action: CloseSnackbarAction) => {
    return {
      notifications: state.notifications.map((notification) =>
        action.dismissAll || notification.key === action.key
          ? { ...notification, dismissed: true }
          : { ...notification },
      ),
    };
  },
  [REMOVE_SNACKBAR]: (state, action: RemoveSnackbarAction) => {
    return {
      notifications: state.notifications.filter((n) => n.key !== action.key),
    };
  },
});

// Selektoren

export const notificationsSelector = (state: RootState): Snackbar[] => state.snackbar.notifications;
