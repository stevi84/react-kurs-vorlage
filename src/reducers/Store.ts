import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { todosReducer } from './TodosReducer';
import { apiCallsReducer } from './ApiCallsReducer';
import { snackbarsReducer } from './SnackbarsReducer';
import { dataTableReducer } from './DataTableReducer';

export const store = configureStore({
  reducer: {
    snackbar: snackbarsReducer,
    todos: todosReducer,
    apiCalls: apiCallsReducer,
    dataTable: dataTableReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
