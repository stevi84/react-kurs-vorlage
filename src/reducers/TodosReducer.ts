import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { createMessage } from '../globals/Translations';
import { Todo } from '../models/Todo';
import * as todoApi from '../apis/TodoApi';
import { AppThunk, RootState } from './Store';
import { decreaseReads, decreaseSubmits, increaseReads, increaseSubmits } from './ApiCallsReducer';
import { enqueueSnackbar } from './SnackbarsReducer';

export const todosInitialState: Todo[] = [];

const todosSlice = createSlice({
  name: 'todos',
  initialState: todosInitialState,
  reducers: {
    setTodos(state, action: PayloadAction<Todo[]>) {
      return action.payload;
    },
    addTodo(state, action: PayloadAction<Todo>) {
      return state.concat(action.payload);
    },
    changeTodo(state, action: PayloadAction<Todo>) {
      return state.map((todo) => (todo.id !== action.payload.id ? todo : action.payload));
    },
    removeTodo(state, action: PayloadAction<number>) {
      return state.filter((todo) => todo.id !== action.payload);
    },
  },
});

const { actions, reducer } = todosSlice;
export const { setTodos, addTodo, changeTodo, removeTodo } = actions;
export { reducer as todosReducer };

export const todosSelector = (state: RootState): Todo[] => state.todos;

export const createTodo =
  (todo: Todo): AppThunk<Promise<Todo>> =>
    async (dispatch) => {
      dispatch(increaseSubmits());
      try {
        const response = await todoApi.createTodo(todo);
        dispatch(addTodo(response.data));
        dispatch(
          enqueueSnackbar(createMessage('create', 'todo', true), { variant: 'success', autoHideDuration: 2000 }),
        );
        return response.data;
      } catch (error) {
        dispatch(enqueueSnackbar(createMessage('create', 'todo', false), { variant: 'error', autoHideDuration: null }));
        throw error;
      } finally {
        dispatch(decreaseSubmits());
      }
    };

export const readTodos = (): AppThunk<Promise<Todo[]>> => async (dispatch) => {
  dispatch(increaseReads());
  try {
    const response = await todoApi.readTodos();
    dispatch(setTodos(response.data));
    return response.data;
  } catch (error) {
    dispatch(enqueueSnackbar(createMessage('read', 'todo', false), { variant: 'error', autoHideDuration: null }));
    throw error;
  } finally {
    dispatch(decreaseReads());
  }
};

export const updateTodo =
  (todo: Partial<Todo>): AppThunk<Promise<Todo>> =>
    async (dispatch) => {
      dispatch(increaseSubmits());
      try {
        const response = await todoApi.updateTodo(todo);
        dispatch(changeTodo(response.data));
        dispatch(
          enqueueSnackbar(createMessage('update', 'todo', true), { variant: 'success', autoHideDuration: 2000 }),
        );
        return response.data;
      } catch (error) {
        dispatch(enqueueSnackbar(createMessage('update', 'todo', false), { variant: 'error', autoHideDuration: null }));
        throw error;
      } finally {
        dispatch(decreaseSubmits());
      }
    };

export const deleteTodo =
  (todoId: number): AppThunk<Promise<void>> =>
    async (dispatch) => {
      dispatch(increaseSubmits());
      try {
        const response = await todoApi.deleteTodo(todoId);
        dispatch(removeTodo(todoId));
        dispatch(
          enqueueSnackbar(createMessage('delete', 'todo', true), { variant: 'success', autoHideDuration: 2000 }),
        );
        return response.data;
      } catch (error) {
        dispatch(enqueueSnackbar(createMessage('delete', 'todo', false), { variant: 'error', autoHideDuration: null }));
        throw error;
      } finally {
        dispatch(decreaseSubmits());
      }
    };
