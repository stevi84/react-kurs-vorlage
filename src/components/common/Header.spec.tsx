import { fireEvent, render, screen } from '@testing-library/react';
import { createRenderer, ShallowRenderer } from 'react-test-renderer/shallow';
import { Header } from './Header';
import { setDataTableVisible } from '../../reducers/DataTableReducer';
import { useDispatch, useSelector } from 'react-redux';

jest.mock('react-redux');
const useSelectorMock = useSelector as jest.Mock;
const useDispatchMock = useDispatch as jest.Mock;

describe('Header', () => {
  it('should equal saved snapshot', () => {
    const renderer: ShallowRenderer = createRenderer();
    renderer.render(<Header />);
    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('should switch logo', () => {
    render(<Header />); 
    expect(screen.queryByTestId('header-logo')).toBeTruthy();
    fireEvent.click(screen.getByLabelText('Logo'));
    expect(screen.queryByTestId('header-logo')).toBeFalsy();
  });

  it('should switch table', () => {
    const mockedDispatch = jest.fn();
    useDispatchMock.mockImplementation(() => mockedDispatch);
    useSelectorMock.mockReturnValue(true);

    render(<Header />);
    fireEvent.click(screen.getByLabelText('Tabelle'));
    expect(mockedDispatch.mock.calls.length).toEqual(1);
    expect(mockedDispatch.mock.calls[0][0]).toEqual(setDataTableVisible(false));
  });
});
