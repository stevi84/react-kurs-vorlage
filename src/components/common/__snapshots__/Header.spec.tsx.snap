// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should equal saved snapshot 1`] = `
<div
  className="header-container"
>
  <div
    className="header-item header-icon"
    data-testid="header-logo"
  >
    üåê
  </div>
  <div
    className="header-item header-text"
  >
    React-App
  </div>
  <div
    className="header-item"
  >
    <input
      checked={true}
      id="header-logo-checkbox"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      htmlFor="header-logo-checkbox"
    >
      Logo
    </label>
  </div>
  <div
    className="header-item"
  >
    <input
      id="header-table-checkbox"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      htmlFor="header-table-checkbox"
    >
      Tabelle
    </label>
  </div>
  <div
    className="header-item"
  >
    <LanguageSwitch />
  </div>
</div>
`;
